packages/main/src/plugin/proxy-resolver.ts

```diff
@@ -0,0 +1,147 @@
+ ***********************************************************************/
+
+import * as http from 'http';
+import * as https from 'https';
+import * as nodeurl from 'url';
+import type { HttpProxyAgentOptions, HttpsProxyAgentOptions } from 'hpagent';
+import { HttpProxyAgent, HttpsProxyAgent } from 'hpagent';
+import type { Proxy } from './proxy.js';
+
+// Agents usage table
+// ------------------------------------
+// | Type            | Proxy | Server |
+// ------------------------------------
+// | HttpProxyAgent  | HTTP  | HTTP   |
+// ------------------------------------
+// | HttpProxyAgent  | HTTPS | HTTP   |
+// ------------------------------------
+// | HttpsProxyAgent | HTTP  | HTTPS  |
+// ------------------------------------
+// | HttpsProxyAgent | HTTPS | HTTPS  |
+// ------------------------------------
+// Source - https://github.com/delvedor/hpagent/tree/main#usage
+
+function createProxyAgent(secure: boolean, proxyUrl: string) {
+  const options = {
+    keepAlive: true,
+    keepAliveMsecs: 1000,
+    maxSockets: 256,
+    maxFreeSockets: 256,
+    scheduling: 'lifo',
+    proxy: proxyUrl,
+  };
+  return secure
+    ? new HttpsProxyAgent(options as HttpProxyAgentOptions)
+    : new HttpProxyAgent(options as HttpsProxyAgentOptions);
+}
+
+function getProxyUrl(proxy: Proxy): string | undefined {
+  if (proxy.isEnabled()) {
+    // use HTTPS proxy if it is configured, because HTTPS proxy works for both HTTP/HTTPS
+    // servers
+    if (proxy.proxy?.httpsProxy) {
+      return proxy.proxy?.httpsProxy;
+    }
+    return proxy.proxy?.httpProxy;
+  }
+}
+
+type ProxyOptions = { agent?: http.Agent | https.Agent };
+
+export function getOptions(proxy: Proxy, secure: boolean): ProxyOptions {
+  const options: ProxyOptions = {};
+  const proxyUrl = getProxyUrl(proxy);
+  if (proxyUrl) {
+    options.agent = createProxyAgent(secure, proxyUrl);
+  }
+  return options;
+}
+
+export function createHttpPatch(originals: typeof http | typeof https, proxy: Proxy) {
+  return {
+    get: patch(originals.get),
+    request: patch(originals.request),
+  };
+
+  function patch(original: typeof http.get) {
+    function patched(
+      url?: string | nodeurl.URL | null,
+      options?: http.RequestOptions | null,
+      callback?: (res: http.IncomingMessage) => void,
+    ): http.ClientRequest {
+      if (proxy?.isEnabled()) {
+        if (typeof url !== 'string' && !url?.searchParams) {
+          callback = <any>options; // eslint-disable-line @typescript-eslint/no-explicit-any
+          options = url;
+          url = undefined;
+        }
+        if (typeof options === 'function') {
+          callback = options;
+          options = undefined;
+        }
+
+        if (!options) {
+          options = {};
+        }
+
+        if (options.socketPath) {
+          return original(options, callback);
+        }
+
+        if (options.agent === true) {
+          throw new Error('Unexpected agent option: true');
+        }
```

Comment:
I see that this code is not tested, but I was wondering when we're reaching that code ?

Reply:
The only invalid value is false according to https://nodejs.org/api/http.html#httprequestoptions-callback. My best guess it is just a check to make client aware of using value, that has no effect.

Reply:
I meant to say `invalid value is 'true'`.